packages/teleport/src/DesktopSession/useDesktopSession.tsx

```diff
@@ -19,23 +19,41 @@ import { useParams } from 'react-router';
 import useAttempt from 'shared/hooks/useAttemptNext';
 import { UrlDesktopParams } from 'teleport/config';
 import Ctx from 'teleport/teleportContext';
-import { useTdpClientCanvas } from './TdpClientCanvas';
+import useTdpClientCanvas from './useTdpClientCanvas';
 
 export default function useDesktopSession(ctx: Ctx) {
-  // Tracks combination of tdpclient/websocket and api call state,
-  // as well as whether the tdp client for this session was intentionally disconnected.
   const { attempt: fetchAttempt, run } = useAttempt('processing');
+
+  // tdpConnection tracks the state of the tdpClient's TDP connection
+  // tdpConnection.status ===
+  // - 'processing' at first
+  // - 'success' once the first TdpClientEvent.IMAGE_FRAGMENT is seen
+  // - 'failed' if a TdpClientEvent.TDP_ERROR is encountered
+  const { attempt: tdpConnection, setAttempt: setTdpConnection } =
```

Comment:
Do we really need this abstraction `tdpConnection` in addition to `wsConnection` ? Why the first `IMAGE_FRAGMENT`  should have any meaning in the context of connection state handling?

Reply:
I think so. The `wsConnection` is instantaneous (barring an incorrect websocket address), but after the websocket connection is made, there is a potentially substantial period of time where the proxy is sending the username and screen size to the windows_service, and then the windows_service is attempting to establish an RDP connection to the windows server, during which we want to display a spinner rather than an empty canvas.

Reply:
> but after the websocket connection is made, there is a potentially substantial period of time where the proxy is sending the username and screen size to the windows_service

I thought that you have fixed it where tdp handshake is done [at websocket connection creation time](https://github.com/gravitational/teleport/blob/master/lib/web/desktop.go#L135). This is why i am asking about `tdpConnection`.

Reply:
The initial handshake is now done in the backend (as you linked to), but it still takes some amount of time for those to be sent from the proxy to the windows_service, and for the windows_service to connect to the windows machine via RDP and receive the first frame.

The proxy is [sending](https://github.com/gravitational/teleport/blob/master/lib/web/desktop.go#L125-L133) those initial TDP messages over the tls connection to the windows_service, and then [blindly proxying](https://github.com/gravitational/teleport/blob/master/lib/web/desktop.go#L135-L168) whatever TDP messages the windows_service sends back. 

The only way we know on the frontend that the RDP connection from the windows_service to the windows host has succeeded is that we get the first `IMAGE_FRAGMENT` back. The websocket connection itself is wrapping TDP, but a successful websocket might yet get back a `TDP_ERROR`.